cloud/pkg/admissioncontroller/controller/admission.go

```diff
@@ -0,0 +1,255 @@
+}
+
+// AdmissionController implements the admission webhook for validation of configuration.
+type AdmissionController struct {
+	Client *kubernetes.Clientset
+}
+
+func strPtr(s string) *string { return &s }
+
+// Register registers the admission webhook.
+// fixme: plugable?
+func (ac *AdmissionController) register(WebhookName string, context *utils.CertContext) error {
+	log.LOGGER.Warn("Get into register")
+	_, err := ac.Client.AdmissionregistrationV1beta1().ValidatingWebhookConfigurations().Create(&admissionregistrationv1beta1.ValidatingWebhookConfiguration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: WebhookName,
+		},
+		Webhooks: []admissionregistrationv1beta1.ValidatingWebhook{
+			{
+				Name: WebhookName,
+				Rules: []admissionregistrationv1beta1.RuleWithOperations{{
+					Operations: []admissionregistrationv1beta1.OperationType{admissionregistrationv1beta1.Create},
+					Rule: admissionregistrationv1beta1.Rule{
+						APIGroups:   []string{"devices.kubeedge.io"},
+						APIVersions: []string{"v1alpha1"},
+						Resources:   []string{"devicemodels"},
+					},
+				}},
+				ClientConfig: admissionregistrationv1beta1.WebhookClientConfig{
+					Service: &admissionregistrationv1beta1.ServiceReference{
+						Namespace: constants.NamespaceName,
+						Name:      constants.ServiceName,
+						Path:      strPtr("/devicemodels"),
+						Port:      pointer.Int32Ptr(constants.Port),
+					},
+					CABundle: context.SigningCert,
+				},
+			},
+		},
+	})
+	log.LOGGER.Warn("End of the register")
+	return err
+}
+
+// Start starts the webhook service
+func (ac *AdmissionController) Start(context *utils.CertContext) {
+	err := ac.register(constants.ExternalAdmissionWebhookName, context)
+	if err != nil {
+		klog.Warningf("Failed to register the webhook with error: %s", err)
+	}
+	ac.deployService()
+	tlsConfig := configTLS(context)
+	http.Handle("/devicemodels", ac) //switch to something like http.HandleFunc("/devicemodels", admitDeviceModel) later.
+	server := &http.Server{
+		Handler:   ac,
+		Addr:      fmt.Sprintf(":%v", constants.Port),
+		TLSConfig: tlsConfig,
+	}
+
+	go func() {
+		if err := server.ListenAndServeTLS("", ""); err != nil {
+			klog.Errorf("ListenAndServeTLS for admission webhook returned error: %v", err)
+		}
+	}()
+}
+
+func (ac *AdmissionController) deployService() {
+	localIP := GetIPv4()
+	if localIP == "" {
+		klog.Error("Cannot get one local valid IP")
+	}
+	//create service
+	service := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: constants.NamespaceName,
+			Name:      constants.ServiceName,
+		},
+		Spec: v1.ServiceSpec{
+			Type: "ClusterIP",
+			Ports: []v1.ServicePort{
+				{
+					Port:       constants.Port,
+					TargetPort: intstr.FromInt(constants.Port),
+				},
+			},
+		},
+	}
+	_, err := ac.Client.CoreV1().Services(constants.NamespaceName).Create(service)
+	if err != nil {
+		klog.Warningf("Failed to create webhook service with error: %s", err)
```

Comment:
should we not abort if the service cannot be created or endpoint cannot be created ? will the admission controller work without these ?

Reply:
Good points, yes, it should abort.

Reply:
rise the log level to `fatal`, so that it will abort, and I thinking a way to opt out this feature so that it won't break other components.